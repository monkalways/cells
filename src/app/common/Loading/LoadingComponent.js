import React from 'react';
import PropTypes from 'prop-types';
import { CircularProgress, withStyles } from '@material-ui/core';

const propTypes = {
  classes: PropTypes.shape({}).isRequired,
  size: PropTypes.number,
};

const defaultProps = {
  size: 100,
};

const LoadingComponent = ({ classes, size }) => (
  <div className={classes.container}>
    <CircularProgress className={classes.progress} size={size} />
  </div>
);

LoadingComponent.propTypes = propTypes;
LoadingComponent.defaultProps = defaultProps;

export default withStyles((theme) => ({
  container: {
    display: 'flex',
    width: '100%',
    height: '100%',
    justifyContent: 'center',
    alignItems: 'center',
  },
  progress: {
    margin: theme.spacing.unit * 2,
  },
}))(LoadingComponent);
