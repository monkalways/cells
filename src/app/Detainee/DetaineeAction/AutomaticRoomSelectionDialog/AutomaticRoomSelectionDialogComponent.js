import React from 'react';
import PropTypes from 'prop-types';
import {
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogTitle,
} from '@material-ui/core';

import Loading from '../../../common/Loading';

const propTypes = {
  areRoomsRefreshing: PropTypes.bool.isRequired,
  currentRoom: PropTypes.string,
  destinationRoom: PropTypes.string,
  detainee: PropTypes.shape({
    firstName: PropTypes.string.isRequired,
    id: PropTypes.string.isRequired,
    lastName: PropTypes.string.isRequired,
    location: PropTypes.string.isRequired,
  }).isRequired,
  getAvailableRoomsRefresh: PropTypes.func.isRequired,
  isAnyRoomAvailable: PropTypes.bool.isRequired,
  isDialogOpen: PropTypes.bool.isRequired,
  isAssigningToRoom: PropTypes.bool.isRequired,
  moveDetaineeToRoom: PropTypes.func.isRequired,
  onClose: PropTypes.func.isRequired,
  usage: PropTypes.string.isRequired,
};

const defaultProps = {
  currentRoom: null,
  destinationRoom: null,
};

const ActivityRoomDialogComponent = ({
  areRoomsRefreshing,
  currentRoom,
  destinationRoom,
  detainee,
  getAvailableRoomsRefresh,
  isAnyRoomAvailable,
  isAssigningToRoom,
  isDialogOpen,
  moveDetaineeToRoom,
  onClose,
  usage,
}) => {
  const handleClick = () => {
    moveDetaineeToRoom(detainee.id, currentRoom, destinationRoom);
  };

  const handleOpen = () => {
    getAvailableRoomsRefresh();
  };

  return (
    <Dialog
      disableBackdropClick={isAssigningToRoom}
      disableEscapeKeyDown={isAssigningToRoom}
      open={isDialogOpen}
      onClose={onClose}
      onRendered={handleOpen}
    >
      {areRoomsRefreshing ? (
        <React.Fragment>
          <DialogTitle>{`Searching for available ${usage} room...`}</DialogTitle>
          <DialogContent>
            <Loading size={50} />
          </DialogContent>
        </React.Fragment>
      ) : isAnyRoomAvailable ? (
        <React.Fragment>
          <DialogTitle>
            {`Moving ${detainee.lastName}, ${
              detainee.firstName
            } to ${usage} room ${destinationRoom}?`}
          </DialogTitle>
          {isAssigningToRoom && (
            <DialogContent>
              <Loading size={50} />
            </DialogContent>
          )}
          <DialogActions>
            <Button
              id="cancelButton"
              onClick={onClose}
              color="primary"
              disabled={isAssigningToRoom}
            >
              Cancel
            </Button>
            <Button
              id="confirmButton"
              onClick={handleClick}
              color="primary"
              variant="contained"
              autoFocus
              disabled={isAssigningToRoom}
            >
              Confirm
            </Button>
          </DialogActions>
        </React.Fragment>
      ) : (
        <React.Fragment>
          <DialogTitle>{`Sorry, the ${usage} room ${destinationRoom} is no longer available.`}</DialogTitle>
          <DialogActions>
            <Button id="closeButton" onClick={onClose} color="primary">
              Close
            </Button>
          </DialogActions>
        </React.Fragment>
      )}
    </Dialog>
  );
};

ActivityRoomDialogComponent.propTypes = propTypes;
ActivityRoomDialogComponent.defaultProps = defaultProps;

export default ActivityRoomDialogComponent;
