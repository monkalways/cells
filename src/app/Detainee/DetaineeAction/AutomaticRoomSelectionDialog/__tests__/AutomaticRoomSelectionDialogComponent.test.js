import React from 'react';
import { shallow } from 'enzyme';
import { Dialog } from '@material-ui/core';

import AutomaticRoomSelectionDialogComponent from '../AutomaticRoomSelectionDialogComponent';
import Loading from '../../../../common/Loading';

describe('ActivityRoomDialogComponent', () => {
  let areRoomsRefreshing;
  let currentRoom;
  let destinationRoom;
  let detainee;
  let getAvailableRoomsRefresh;
  let isAnyRoomAvailable;
  let isDialogOpen;
  let isAssigningToRoom;
  let moveDetaineeToRoom;
  let onClose;
  let usage;

  beforeEach(() => {
    areRoomsRefreshing = false;
    currentRoom = 'C1';
    destinationRoom = 'T4';
    detainee = {
      firstName: 'Bob',
      id: '123',
      lastName: 'Saget',
      location: 'location',
    };
    getAvailableRoomsRefresh = jest.fn();
    isAnyRoomAvailable = false;
    isDialogOpen = false;
    isAssigningToRoom = false;
    moveDetaineeToRoom = jest.fn();
    onClose = jest.fn();
    usage = 'usage';
  });

  const setup = () => shallow(<AutomaticRoomSelectionDialogComponent
    areRoomsRefreshing={areRoomsRefreshing}
    currentRoom={currentRoom}
    destinationRoom={destinationRoom}
    detainee={detainee}
    getAvailableRoomsRefresh={getAvailableRoomsRefresh}
    isAnyRoomAvailable={isAnyRoomAvailable}
    isDialogOpen={isDialogOpen}
    isAssigningToRoom={isAssigningToRoom}
    moveDetaineeToRoom={moveDetaineeToRoom}
    onClose={onClose}
    usage={usage}
  />);

  it('should render Loading component if activity rooms are refreshing', () => {
    areRoomsRefreshing = true;

    const wrapper = setup();
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(Loading)).toExist();
  });

  it('should show a close button when the room is no longer available', () => {
    areRoomsRefreshing = false;
    isAnyRoomAvailable = false;

    const wrapper = setup();
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('#closeButton')).toExist();
  });

  it('should show confirm and cancel buttons when moving a detainee', () => {
    areRoomsRefreshing = false;
    isAnyRoomAvailable = true;
    isAssigningToRoom = false;

    const wrapper = setup();
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('#cancelButton')).toExist();
    expect(wrapper.find('#confirmButton')).toExist();
    expect(wrapper.find(Loading)).not.toExist();
  });

  it('should show a loading component when it is assigning the detainee to a room', () => {
    areRoomsRefreshing = false;
    isAnyRoomAvailable = true;
    isAssigningToRoom = true;

    const wrapper = setup();
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('#cancelButton')).toExist();
    expect(wrapper.find('#confirmButton')).toExist();
    expect(wrapper.find(Loading)).toExist();
  });

  it('should handle the dialog closing', () => {
    const wrapper = setup();
    wrapper.find(Dialog).simulate('close');
    expect(onClose).toBeCalled();
  });

  it('should handle close button click', () => {
    areRoomsRefreshing = false;
    isAnyRoomAvailable = false;

    const wrapper = setup();
    wrapper.find('#closeButton').simulate('click');
    expect(onClose).toBeCalled();
  });

  it('should handle confirm button click', () => {
    areRoomsRefreshing = false;
    isAnyRoomAvailable = true;

    const wrapper = setup();
    wrapper.find('#confirmButton').simulate('click');
    expect(moveDetaineeToRoom).toBeCalledWith(
      detainee.id,
      currentRoom,
      destinationRoom,
    );
  });

  it('should handle cancel button click', () => {
    areRoomsRefreshing = false;
    isAnyRoomAvailable = true;

    const wrapper = setup();
    wrapper.find('#cancelButton').simulate('click');
    expect(onClose).toBeCalled();
  });
});
