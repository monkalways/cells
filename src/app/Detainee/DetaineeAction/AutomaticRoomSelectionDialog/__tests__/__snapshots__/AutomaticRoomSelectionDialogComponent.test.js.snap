// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActivityRoomDialogComponent should render Loading component if activity rooms are refreshing 1`] = `
<WithStyles(Dialog)
  disableBackdropClick={false}
  disableEscapeKeyDown={false}
  onClose={[MockFunction]}
  onRendered={[Function]}
  open={false}
>
  <WithStyles(DialogTitle)>
    Searching for available usage room...
  </WithStyles(DialogTitle)>
  <WithStyles(DialogContent)>
    <WithStyles(LoadingComponent)
      size={50}
    />
  </WithStyles(DialogContent)>
</WithStyles(Dialog)>
`;

exports[`ActivityRoomDialogComponent should show a close button when the room is no longer available 1`] = `
<WithStyles(Dialog)
  disableBackdropClick={false}
  disableEscapeKeyDown={false}
  onClose={[MockFunction]}
  onRendered={[Function]}
  open={false}
>
  <WithStyles(DialogTitle)>
    Sorry, the usage room T4 is no longer available.
  </WithStyles(DialogTitle)>
  <WithStyles(DialogActions)>
    <WithStyles(Button)
      color="primary"
      id="closeButton"
      onClick={[MockFunction]}
    >
      Close
    </WithStyles(Button)>
  </WithStyles(DialogActions)>
</WithStyles(Dialog)>
`;

exports[`ActivityRoomDialogComponent should show a loading component when it is assigning the detainee to a room 1`] = `
<WithStyles(Dialog)
  disableBackdropClick={true}
  disableEscapeKeyDown={true}
  onClose={[MockFunction]}
  onRendered={[Function]}
  open={false}
>
  <WithStyles(DialogTitle)>
    Moving Saget, Bob to usage room T4?
  </WithStyles(DialogTitle)>
  <WithStyles(DialogContent)>
    <WithStyles(LoadingComponent)
      size={50}
    />
  </WithStyles(DialogContent)>
  <WithStyles(DialogActions)>
    <WithStyles(Button)
      color="primary"
      disabled={true}
      id="cancelButton"
      onClick={[MockFunction]}
    >
      Cancel
    </WithStyles(Button)>
    <WithStyles(Button)
      autoFocus={true}
      color="primary"
      disabled={true}
      id="confirmButton"
      onClick={[Function]}
      variant="contained"
    >
      Confirm
    </WithStyles(Button)>
  </WithStyles(DialogActions)>
</WithStyles(Dialog)>
`;

exports[`ActivityRoomDialogComponent should show confirm and cancel buttons when moving a detainee 1`] = `
<WithStyles(Dialog)
  disableBackdropClick={false}
  disableEscapeKeyDown={false}
  onClose={[MockFunction]}
  onRendered={[Function]}
  open={false}
>
  <WithStyles(DialogTitle)>
    Moving Saget, Bob to usage room T4?
  </WithStyles(DialogTitle)>
  <WithStyles(DialogActions)>
    <WithStyles(Button)
      color="primary"
      disabled={false}
      id="cancelButton"
      onClick={[MockFunction]}
    >
      Cancel
    </WithStyles(Button)>
    <WithStyles(Button)
      autoFocus={true}
      color="primary"
      disabled={false}
      id="confirmButton"
      onClick={[Function]}
      variant="contained"
    >
      Confirm
    </WithStyles(Button)>
  </WithStyles(DialogActions)>
</WithStyles(Dialog)>
`;
